File: front/projects/web/src/types/query.ts
import { ParseNumber } from "@/utils"

export interface QueryInfo {
  keywords: string
  mode: string
  filter: {
    [key: string]: any
  }
  pageInfo?: PageInfo
  sortInfo?: SortInfo
}

export default class Query {
  keywords: string
  mode: string
  filter: {
    [key: string]: any
  }
  pageInfo: PageInfo
  sortInfo: SortInfo

  constructor(queryInfo: any) {
    this.keywords = queryInfo["keywords"] ? queryInfo["keywords"] : ""
    this.mode = queryInfo["mode"] ? queryInfo["mode"] : ""
    this.filter = {}
    this.pageInfo = {
      index: ParseNumber(queryInfo["pageIndex"], 1)!,
      size: ParseNumber(queryInfo["pageIndex"], 20)!
    }
    this.sortInfo = {
      field: queryInfo["field"] ? queryInfo["field"] : "",
      order: ParseQueryOrder(queryInfo["order"])
    }
  }
}

export interface PageInfo {
  index: number
  size: number
}

export function NewPageInfo(index: number, size: number): PageInfo {
  return {
    index: index,
    size: size
  }
}

export interface SortInfo {
  field: string
  order: "asc" | "desc" | ""
}

export function NewSortInfo(field: string, order: "asc" | "desc" | ""): SortInfo {
  return {
    field: field,
    order: order
  }
}

export interface QueryList<T> {
  total: number
  pageInfo: PageInfo
  objects: T[]
}

function ParseQueryOrder(value: any, defaultValue: "desc" | "asc" = "asc") {
  const result = toString(value).trim()
  if (["desc", "asc"].indexOf(result) === -1) {
    return defaultValue
  }
  return result
}

File: backend/types/query.go
package types

type QueryInfo struct {
	Keywords string         `json:"keywords"`
	Mode     string         `json:"mode"`
	Filter   map[string]any `json:"filter"`
	PageInfo *PageInfo      `json:"pageInfo"`
	SortInfo *SortInfo      `json:"sortInfo"`
}

func (q *QueryInfo) CheckBase() {
	if q.Filter == nil {
		q.Filter = make(map[string]any)
	}
	if q.PageInfo == nil {
		if q.PageInfo.Index < 1 {
			q.PageInfo.Index = 1
		}
		if q.PageInfo.Size < 1 {
			q.PageInfo.Size = 20
		}
	}
	if q.SortInfo == nil {
		if q.SortInfo.Field == "" {
			q.SortInfo = nil
		} else if q.SortInfo.Order != "asc" && q.SortInfo.Order != "desc" {
			q.SortInfo.Order = "asc"
		}
	}
}

type SortInfo struct {
	Field string `json:"field"`
	Order string `json:"order"`
}

type PageInfo struct {
	Index int `json:"index"`
	Size  int `json:"size"`
}

type QueryResult[T any] struct {
	Total int  `json:"total"`
	Data  []*T `json:"data"`
}
